core/capabilities/ccip: fix oracle startup logic

makramkd opened the pull request:
Ticket: https://smartcontract-it.atlassian.net/browse/CCIP-3078

We can't start bootstrap oracles and plugin oracles on the same chainlink node. The main reason for this is that we have a singleton `SingletonPeerWrapper` object which manages all libocr peers for the application as a whole. Since this singleton only works with a single peer ID, it creates a single OCR peer, and all streams are created using this OCR peer. Since streams must have unique config digests for the same peer object, running a bootstrap oracle and a plugin oracle for the same config digest will fail because the stream will not be created. 

In order to remedy this we match what will be the case in production, which is:

* Separate bootstrap node, with a peer ID that is either part of the OCR committee or not (most likely the latter, in order to avoid exporting / importing the P2P key from one node's DB to another). This bootstrap node will have to have a DNS name and be publicly accessible over the internet in order to be accessed by all nodes in the committee, at least initially, to facilitate peer discovery.
* Plugin node that is more locked down, i.e no public DNS name required.

To enable this, in this PR we:

* Tweak the `launcher` component to launch bootstrap nodes only or plugin nodes only, and not both. This does NOT rely on the on-chain bootstrap P2P IDs that are set in the OCR config. This field will be removed in subsequent PRs.
* Tweak the `OracleCreator` interface to support the above operation

Follow ups:
* Remove the `bootstrapP2PIds` field from the OCR config in CCIPConfig.sol

makramkd committed:
core/capabilities/ccip: fix oracle startup logic

Ticket: https://smartcontract-it.atlassian.net/browse/CCIP-3078

Remove the bootstrap oracle startup logic. This is
actually not needed since we specify V2Bootstrappers
on the libocr3.OracleArgs struct, since this will
cause the created oracle to automatically bootstrap
without the need for the bootstrap node.

This does mean that it is vital that the P2PV2Bootstrappers
field in the CCIP capability job spec is correctly set.

This also speeds up tests mildly by about 20% which
is a good quality of life bonus.

makramkd committed:
run mockery

makramkd committed:
goimports

makramkd committed:
launcher with different modes

makramkd committed:
revert comment change

makramkd committed:
simplify oracle creator interface

makramkd committed:
launch bootstrap node in test

makramkd committed:
revert comment

connorwstein commented on contracts/src/v0.8/ccip/capability/CCIPConfig.sol:
```diff
@@ -393,11 +393,7 @@ contract CCIPConfig is ITypeAndVersion, ICapabilityConfiguration, OwnerIsCreator
     if (cfg.p2pIds.length != cfg.signers.length || cfg.p2pIds.length != cfg.transmitters.length) {
       revert P2PIdsLengthNotMatching(cfg.p2pIds.length, cfg.signers.length, cfg.transmitters.length);
     }
-    if (cfg.bootstrapP2PIds.length > cfg.p2pIds.length) revert TooManyBootstrapP2PIds();
```
just because theoretically could have more?

connorwstein commented on contracts/src/v0.8/ccip/capability/libraries/CCIPConfigTypes.sol:
```diff
@@ -37,7 +37,6 @@ library CCIPConfigTypes {
     uint64 offchainConfigVersion; // ─────────────╯ The version of the offchain configuration.
     bytes offrampAddress; // The remote chain offramp address.
-    // NOTE: bootstrapP2PIds and p2pIds should be sent as sorted sets
     bytes32[] bootstrapP2PIds; // The bootstrap P2P IDs of the oracles that are part of the role DON.
```
hmm do we even need bootstrapP2PIds onchain now? You _could_ have an offchain assertion that the job spec bootstrapP2PIds match, but not sure that's really valuable?

makramkd replied:
We don't - I'm gonna make the change to remove them in a separate PR to unblock this fix. Will revert the contract changes in this PR so as to not increase the scope massively, but removing them is a breaking change that will require a change in chainlink-ccip first.

connorwstein commented on core/capabilities/ccip/validate/validate.go:
```diff
@@ -40,9 +40,6 @@ func ValidatedCCIPSpec(tomlString string) (jb job.Job, err error) {
 	if jb.CCIPSpec.P2PKeyID == "" {
 		return job.Job{}, fmt.Errorf("p2pKeyID must be set")
 	}
-	if len(jb.CCIPSpec.P2PV2Bootstrappers) == 0 {
```
would it make sense to ensure that if len(jb.CCIPSpec.P2PV2Bootstrappers) == 0, then there are no OCR bundles or relays other than homechain config set?

connorwstein replied:
I also noticed that we don't use the fallback to env vars like in V1 
```	bootstrapPeers, err := ocrcommon.GetValidatedBootstrapPeers(spec.P2PV2Bootstrappers, d.peerWrapper.P2PConfig().V2().DefaultBootstrappers())
	if err != nil {
		return nil, err
	}
```
I think only really makes sense in the "huge number of job specs don't want to repeat" case so makes sense to remove, afaict its not required config. But we should make it clear to NOPs/example TOMLs that the P2PV2Default bootstrappers would not be used for ccip cap jobs.

makramkd replied:
> would it make sense to ensure that if len(jb.CCIPSpec.P2PV2Bootstrappers) == 0, then there are no OCR bundles or relays other than homechain config set?

It would, but we can't validate that solely from this ValidatedCCIPSpec func which only gets the spec. We can have this check in the `ServicesForSpec` func on the delegate.

> But we should make it clear to NOPs/example TOMLs that the P2PV2Default bootstrappers would not be used for ccip cap jobs.

Yeah I figured since we'd be pushing the job specs to them and that the p2pID is part of the spec, it would be impossible for the nops to even affect that with a busted TOML config.

makramkd replied:
> It would, but we can't validate that solely from this ValidatedCCIPSpec func which only gets the spec. We can have this check in the ServicesForSpec func on the delegate.

Actually I take it back, I wonder what the usefulness of this validation would be

connorwstein replied:
I think just to make a clear difference between bootstrap jobs and non-bootstraps. Confusing to look at a running job and see it has keys but is only bootstrapping

makramkd replied:
OK yeah I think thats fine, I can add that validation.

makramkd replied:
Will fix in an upcoming PR, integration test is failing after adding the validations for strange reasons. Merging this in to at least have a fix and unblock updating CCIPConfig

makramkd replied:
https://smartcontract-it.atlassian.net/browse/CCIP-3123

makramkd committed:
revert changes to ccipconfig

makramkd committed:
feedback: don't use bootstrap p2p ids from onchain

Don't use the bootstrapP2PIds set onchain in the OCR config, instead,
the CCIP capability job will run in either "bootstrap mode" or "plugin
mode" depending on the P2PV2Bootstrappers provided to the job spec.

makramkd committed:
fix integration-tests

makramkd committed:
Merge branch 'ccip-develop' into mk/ccip-3078/no-bootstrap

dimkouv commented on core/capabilities/ccip/delegate.go:
```diff
@@ -134,30 +135,45 @@ func (d *Delegate) ServicesForSpec(ctx context.Context, spec job.Job) (services
-		bootstrapperLocators,
-		hcr,
-	)
+	// if bootstrappers are provided we assume that the node is a plugin oracle.
```
nit: 
```
isPluginOracle = len(spec.CCIPSpec.P2PV2Bootstrappers) > 0
if isPluginOracle { ... }
```

makramkd committed:
goimports

dimkouv commented on core/capabilities/ccip/oraclecreator/bootstrap.go:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	destChainFamily := chaintype.EVM
+	destRelayID := types.NewRelayID(string(destChainFamily), fmt.Sprintf("%d", chainID))
```
nit: use `strconv` instead of `fmt.Sprintf`

makramkd committed:
Merge branch 'ccip-develop' into mk/ccip-3078/no-bootstrap

makramkd committed:
Merge branch 'ccip-develop' into mk/ccip-3078/no-bootstrap

makramkd committed:
Merge branch 'ccip-develop' into mk/ccip-3078/no-bootstrap
